alf/algorithms/handcrafted_algorithm.py

```diff
@@ -0,0 +1,150 @@
+                provided to the algorithm which performs ``train_iter()`` by
+                itself.
+            debug_summaries (bool): True if debug summaries should be created.
+            name (str): The name of this algorithm.
+        """
+
+        super().__init__(
+            observation_spec,
+            action_spec,
+            train_state_spec=(),
+            env=env,
+            config=config,
+            debug_summaries=debug_summaries,
+            name=name)
+
+    def _policy_func(self, observation):
+        """A function calculating action based on the input observation.
+        Each subclass needs to define this function
+
+        Args:
+            observation (nested Tensor): input observation that is compatible
+                with observation_spec
+        Returns:
+            nested Tensor: action that is compatible with action spec
+        """
+        raise NotImplementedError('Must define _policy_func member '
+                                  'function for the class')
+
+    def _predict_action(self, observation, state):
+        """Predict action based on observation
+        """
+        return self._policy_func(observation)
+
+    def predict_step(self, time_step: TimeStep, state, epsilon_greedy=1.0):
+        action = self._predict_action(time_step.observation, state=state)
+        return AlgStep(output=action, state=state)
+
+    def rollout_step(self, time_step: TimeStep, state):
+        action = self._predict_action(time_step.observation, state=state)
+        return AlgStep(output=action, state=state)
+
+    def train_step(self, exp: Experience, state):
+        return AlgStep()
+
+    def calc_loss(self, experience, train_info):
+        return LossInfo()
+
+
+@gin.configurable
+class SimpleCarlaAlgorithm(HandcraftedAlgorithm):
+    """A simple controller for Carla environment.
+    """
+
+    def __init__(self,
+                 observation_spec,
+                 action_spec: BoundedTensorSpec,
+                 env=None,
+                 config: TrainerConfig = None,
+                 debug_summaries=False,
+                 name="SimpleCarlaAlgorithm"):
+
+        super().__init__(
+            observation_spec,
+            action_spec,
+            env=env,
+            config=config,
+            debug_summaries=debug_summaries,
+            name=name)
+
+    def _policy_func(self, observation):
+        """A naive hand-crafted policy for Carla environment.
+
+        Args:
+            observation (nested Tensor): input observation that is compatible
+                with observation_spec
+        Returns
+            Tensor: action computed based on the observation
+        """
+        nav = alf.nest.get_field(observation, 'observation.navigation')
+        goal = alf.nest.get_field(observation, 'observation.goal')
+
+        waypoints = nav
+        if waypoints.shape[-1] > 1:
+            wp_vector = waypoints[:, 1]
+        else:
+            wp_vector = waypoints[:, 0]
```

Comment:
The shape of waypoints is [B, k, 3], where B is the batch size, k is the number of way points.
Should the if statement be `if waypoints.shape[1] > 1`?

Reply:
Bug fixed.